src/phylo_flags.hpp

```diff
@@ -0,0 +1,336 @@
+class PhyloMapkey {
+ public:
+  PhyloMapkey(const std::string &name, const std::string &key);
+
+  // Comparators
+  static int Compare(const PhyloMapkey &mapkey_a, const PhyloMapkey &mapkey_b);
+  // General compare.
+  bool operator==(const PhyloMapkey &other) const;
+  friend bool operator==(const PhyloMapkey &lhs, const PhyloMapkey &rhs);
+  bool operator<(const PhyloMapkey &other) const;
+  friend bool operator<(const PhyloMapkey &lhs, const PhyloMapkey &rhs);
+  // Compare against String
+  bool operator==(const std::string &other_name) const;
+  bool operator<(const std::string &other_name) const;
+
+  // Getters
+  std::string GetName() const { return name_; };
+  std::string GetKey() const { return key_; };
+  std::string operator()() const { return GetKey(); };
+
+ private:
+  // This is a descriptive name of the mapkey that will be visible to the user in
+  // bito pybind interface.
+  std::string name_;
+  // This is the uniquely identifiable key that is used for accessing a map location.
+  // For convenience, this key should match the input flag if that flag
+  // directly corresponds to a function mapkey for the computation of underlying data.
+  // (e.g. in FatBeagle::Gradient, we have an mapkey to compute
+  // `substitution_model_rates`. If flag is set, then the return map will contain a
+  // `substitution_model_rates` key).
+  std::string key_;
+};
+
+// ** Phylo Mapkey Set
+// Contains all possible options for function.
+class PhyloMapkeySet {
+ public:
+  using MapkeyMap = std::map<std::string, PhyloMapkey>;
+
+  explicit PhyloMapkeySet(const std::string &name) : name_(name){};
+  PhyloMapkeySet(const std::string &name, const std::vector<PhyloMapkey> &mapkeys);
+
+  // Insert individual mapkey.
+  void AddMapkey(const PhyloMapkey &mapkey, const bool overwrite = false);
+  // Does mapkey already exist in set?
+  bool ContainsMapkey(const PhyloMapkey &mapkey);
+  const MapkeyMap &GetMapkeys() const { return all_mapkeys_; };
+  std::string ToString() const;
+
+ private:
+  // Name for mapkey set.
+  std::string name_;
+  // List of all possible keys.
+  MapkeyMap all_mapkeys_;
+};
+
+// ** Phylo FlagOption
+// This is the base option flag type.  Also specifies default behaviour for flags.
+class PhyloFlagOption {
+ public:
+  enum class FlagType { None, Include, Exclude, SetValue, MapKey, RunAll };
+  enum class DataType { None, Bool, Double, String };
+
+  PhyloFlagOption();
+  PhyloFlagOption(const std::string &name, const std::string &flag,
+                  const bool is_set_when_running_defaults, const FlagType flag_type,
+                  const DataType data_type);
+  // PhyloFlagOption FlagType specific constructors.
+  static PhyloFlagOption IncludeOption(const std::string &name, const std::string &flag,
+                                       const bool is_set_when_running_defaults = true);
+  static PhyloFlagOption ExcludeOption(const std::string &name, const std::string &flag,
```

Comment:
For discussion: compare to https://click.palletsprojects.com/en/8.0.x/options/#boolean-flags

Reply:
Let's remove

Reply:
Combined `IncludeOption` and `ExcludeOption` into just `BooleanOption`